src/qibolab/instruments/qblox/acquisition.py

```diff
@@ -0,0 +1,146 @@
+from dataclasses import dataclass
+from typing import List, Optional, Tuple
+
+import numpy as np
+
+SAMPLING_RATE = 1e9
+
+
+def demodulate(input_i, input_q, frequency):
+    """Demodulates and integrates the acquired pulse."""
+    # DOWN Conversion
+    # qblox does not remove the offsets in hardware
+    modulated_i = input_i - np.mean(input_i)
+    modulated_q = input_q - np.mean(input_q)
+
+    num_samples = modulated_i.shape[0]
+    time = np.arange(num_samples) / SAMPLING_RATE
+    cosalpha = np.cos(2 * np.pi * frequency * time)
+    sinalpha = np.sin(2 * np.pi * frequency * time)
+    demod_matrix = np.sqrt(2) * np.array([[cosalpha, sinalpha], [-sinalpha, cosalpha]])
+    result = []
+    for it, t, ii, qq in zip(np.arange(modulated_i.shape[0]), time, modulated_i, modulated_q):
+        result.append(demod_matrix[:, :, it] @ np.array([ii, qq]))
+    return np.mean(np.array(result), axis=0)
+
+
+@dataclass
+class AveragedAcquisition:
+    """Software Demodulation.
+
+    Every readout pulse triggers an acquisition, where the 16384 i and q samples of the waveform
+    acquired by the ADC are saved into a dedicated memory within the FPGA. This is what qblox calls
+    *scoped acquisition*. The results of multiple shots are averaged in this memory, and cannot be
+    retrieved independently. The resulting waveforms averages (i and q) are then demodulated and
+    integrated in software (and finally divided by the number of samples).
+    Since Software Demodulation relies on the data of the scoped acquisition and that data is the
+    average of all acquisitions, **only one readout pulse per qubit is supported**, so that
+    the averages all correspond to reading the same quantum state.
+    """
+
+    raw: Tuple[List[float], List[float]]
+    """Tuples with the averages of the i and q waveforms for every readout pulse ([i samples], [q samples])."""
+
+    demodulated_integrated: Tuple[float, float, float, float]
+    """Tuples with the results of demodulating and integrating (averaging over time) the average of the
+    waveforms for every pulse: ``(amplitude[V], phase[rad], i[V], q[V])``."""
+
+    @property
+    def data(self):
+        """Acquisition data to be returned to the platform.
+
+        Ignores the data available in acquisition results and returns only i and q voltages.
+        """
+        # TODO: to be updated once the functionality of ExecutionResults is extended
+        return self.demodulated_integrated[2:]  # (i, q)
+
+    @property
+    def default(self):
+        """Default Results: Averaged Demodulated Integrated"""
+        return self.demodulated_integrated
+
+    @classmethod
+    def create(cls, data, duration, frequency):
+        scope = data["acquisition"]["scope"]
+        raw = (
+            scope["path0"]["data"][0:duration],
+            scope["path1"]["data"][0:duration],
+        )
+        i, q = demodulate(*raw, frequency)
```

Comment:
Since this is a possibly expensive operation, it would make sense to postpone it (as in the case of the `@cachedproperty`).
Just adopt a lazy approach, rather than an eager one.